### PR TITLE
Added derivation script for co2s

### LINKED ISSUE #586: Add derivation script for co2 mole fraction at surface
The variable `co2` (`mole_fraction_of_carbon_dioxide_in_air`) is 3D. I need a 2D version of this variable which gives the CO2 mole fraction at surface level.

For unmasked datasets this is trivial (just use the lowest pressure level), but for masked this is a little bit more tricky (extract the lowest value which is not masked for every grid cell).

I will open a PR for this.

### PR DESCRIPTION
**Pull request by schlunma**
This PR adds the derivation script for `co2s` (CO2 mole fraction at surface level). It can be tested with this sample recipe:

```yaml
# recipe_test.yml
---
documentation:

  description: |
    Test.

  authors:
    - schlund_manuel

  maintainer:
    - schlund_manuel

  projects:
    - crescendo


DATASET_ANCHOR: &datasets
  - {dataset: BNU-ESM}
  - {dataset: CanESM2}
  - {dataset: CESM1-BGC}
  - {dataset: GFDL-ESM2G}
  - {dataset: GFDL-ESM2M}
  - {dataset: MIROC-ESM}
  - {dataset: NorESM1-ME}


diagnostics:

  test:
    variables:
      co2s:
        project: CMIP5
        exp: esmHistorical
        ensemble: r1i1p1
        mip: Amon
        additional_datasets: *datasets
        start_year: 1950
        end_year: 1955
        derive: true
    scripts:
      null
```

Please note that using `core_data()` instead of `data` [here](https://github.com/ESMValGroup/ESMValCore/blob/dede748b065741e76590982ca0e1c5e5b94a7a40/esmvalcore/preprocessor/_derive/co2s.py#L35) results in an `NotImplementedError: Don't yet support nd fancy indexing` of `dask`.

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #586.

### CODE REVIEW

**Comment by schlunma** [comment]
I think the fix for `co2` of CESM1-BGC is wrong:

https://github.com/ESMValGroup/ESMValCore/blob/735f4160da4d2ae113dd87f93f25cb38d943c656/esmvalcore/cmor/_fixes/cmip5/cesm1_bgc.py#L25-L28

For the year 1950, the resulting CO2 fraction is ~200ppm, while all other models have around 300ppm. When I compare the original file to that of BNU-ESM (which has the same fix), BNU-ESM shows CO2 fractions of over 500ppm in 2005, while `CESM1-BGC` has about 400ppm (`historical` and `esmHistorical` exp), which is the correct value. Thus, I think that BNU-ESM needs that fix, but CESM1-BGC does not.

Can someone with access to mistral confirm that? These are the files

`/pf/b/b309141/work/CMIP5_DKRZ/BNU/BNU-ESM/esmHistorical/mon/atmos/Amon/r1i1p1/v20120510/co2/co2_Amon_BNU-ESM_esmHistorical_r1i1p1_185001-200512.nc`

`/pf/b/b309141/work/CMIP5_DKRZ/NSF-DOE-NCAR/CESM1-BGC/esmHistorical/mon/atmos/Amon/r1i1p1/v20130216/co2/co2_Amon_CESM1-BGC_esmHistorical_r1i1p1_195001-200512.nc`

**Comment by mattiarighi** [comment]
I also think the fix is wrong.

The multiplication factor `molarmass(C) / molarmass(CO2)` is usually applied to mass units, but here it is applied to a volume mixing ratio.

**Comment by mattiarighi** [comment]
It is actually molarmass(dry_air) / molarmass(co2).

File: esmvalcore/cmor/tables/custom/CMOR_co2s.dat [hunk 1]
```diff
@@ -0,0 +1,20 @@
+SOURCE: CMIP5
+!============
+variable_entry:    co2s
+!============
+modeling_realm:    atmos
+!----------------------------------
+! Variable attributes:
+!----------------------------------
+standard_name:     mole_fraction_of_carbon_dioxide_in_air
+units:             mol mol-1
```
Comment (mattiarighi) [inline esmvalcore/cmor/tables/custom/CMOR_co2s.dat]: Is there a reason for not using the same units as the parent variable (`1.e-6`)?

Comment (schlunma) [reply esmvalcore/cmor/tables/custom/CMOR_co2s.dat]: I used the one from the CMIP6 table:

https://github.com/ESMValGroup/ESMValCore/blob/735f4160da4d2ae113dd87f93f25cb38d943c656/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Amon.json#L305-L321

Comment (mattiarighi) [reply esmvalcore/cmor/tables/custom/CMOR_co2s.dat]: Ok. I was looking into CMIP5.

**Comment by schlunma** [comment]
@mattiarighi I just realized that there is already a variable `co2s` in the CMIP6 tables :grimacing: :

https://github.com/ESMValGroup/ESMValCore/blob/cd5cf1f622fd277323008c054a1f19f54f73de64/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_Emon.json#L827-L844

Unfortunately, this can't be used for all projects which use `CMIP5`-like tables, but I need to use that variable with CMIP5. What should we do here? I see two options:

1. Rename the custom `co2s`

2. Change the custom `co2s` so that it is similar to the CMIP6 entry.

What do you think?

**Comment by mattiarighi** [comment]
I would go for the second option, using two names for the same variables sounds confusing.
But we should discuss this in the @ESMValGroup/esmvaltool-coreteam, since I think we never had this case before, did we?

**Comment by valeriupredoi** [comment]
You can map CMIP6<->CMIP5 variables as @jvegasbsc implemented in the cmor module, then you can add this variable as a CMIP5 variable only and map it to the actual cmor tables CMIP6 variable. Hows that sound? :beer:

**Comment by schlunma** [comment]
@valeriupredoi can you show me where this CMIP5<->CMIP6 translation is in the code?

**Comment by valeriupredoi** [comment]
`l.90` in `esmvalcore/cmor/table.py`: `_CMIP_5to6_varname` dict :beer:

**Comment by schlunma** [comment]
But this is only necessary if the the variables have different names in CMIP5 and CMIP6, right? If we use the second option, the variable is called `co2s` in both cases.

**Comment by jvegasbsc** [comment]
Yes, you need to do this only for those that differ in both

**Comment by valeriupredoi** [comment]
yeh I was mistaken in thinking it needs a different name for CMIP5 :beer:

**Pull request merged**